diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4f3346f..fe2ed9a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1861,7 +1861,16 @@ if(NOT CURL_DISABLE_INSTALL)
     )
   endif()
 
-  install(FILES ${version_config} ${project_config}
+  set(components_file "${CMAKE_CURRENT_BINARY_DIR}/CURLConfigComponents.cmake")
+  file(CONFIGURE OUTPUT "${components_file}" CONTENT [[
+  foreach(z_vcpkg_curl_component IN ITEMS @SUPPORT_FEATURES@ @SUPPORT_PROTOCOLS@)
+    if(z_vcpkg_curl_component MATCHES "^[-_a-zA-Z0-9]*$")
+      set(CURL_${z_vcpkg_curl_component}_FOUND TRUE)
+    endif()
+  endforeach()
+  ]] @ONLY)
+
+  install(FILES ${version_config} ${project_config} ${components_file}
           DESTINATION ${CURL_INSTALL_CMAKE_DIR}
   )
 
