diff --git a/include/curl/curl.h b/include/curl/curl.h
index 9e07527dc9..2210150ea6 100644
--- a/include/curl/curl.h
+++ b/include/curl/curl.h
@@ -960,6 +960,9 @@ typedef enum {
 /* If possible, send data using TLS 1.3 early data */
 #define CURLSSLOPT_EARLYDATA (1L<<6)
 
+#define CURLSSLOPT_TLSEXT_PADDING (1<<7)
+#define CURLSSLOPT_TLSEXT_PADDING_SUPER (1<<8)
+
 /* The default connection attempt delay in milliseconds for happy eyeballs.
    CURLOPT_HAPPY_EYEBALLS_TIMEOUT_MS.3 and happy-eyeballs-timeout-ms.d document
    this value, keep them in sync. */
diff --git a/lib/setopt.c b/lib/setopt.c
index 3c3adb06a9..3b6f225697 100644
--- a/lib/setopt.c
+++ b/lib/setopt.c
@@ -435,6 +435,8 @@ static void set_ssl_options(struct ssl_config_data *ssl,
   ssl->native_ca_store = !!(arg & CURLSSLOPT_NATIVE_CA);
   ssl->auto_client_cert = !!(arg & CURLSSLOPT_AUTO_CLIENT_CERT);
   ssl->earlydata = !!(arg & CURLSSLOPT_EARLYDATA);
+  ssl->tlsext_padding = !!(arg & CURLSSLOPT_TLSEXT_PADDING);
+  ssl->tlsext_padding_super = !!(arg & CURLSSLOPT_TLSEXT_PADDING_SUPER);
 }
 #endif
 
diff --git a/lib/urldata.h b/lib/urldata.h
index f79ccca0ac..3b0c087c27 100644
--- a/lib/urldata.h
+++ b/lib/urldata.h
@@ -283,6 +283,8 @@ struct ssl_config_data {
   BIT(custom_cafile); /* application has set custom CA file */
   BIT(custom_capath); /* application has set custom CA path */
   BIT(custom_cablob); /* application has set custom CA blob */
+  BIT(tlsext_padding);
+  BIT(tlsext_padding_super);
 };
 
 struct ssl_general_config {
diff --git a/lib/vtls/openssl.c b/lib/vtls/openssl.c
index 764d829325..69b1c68eac 100644
--- a/lib/vtls/openssl.c
+++ b/lib/vtls/openssl.c
@@ -4180,6 +4235,14 @@ CURLcode Curl_ossl_ctx_init(struct ossl_ctx *octx,
   if(!ssl_config->enable_beast)
     ctx_options &= ~(ctx_option_t)SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS;
 
+  if(ssl_config->tlsext_padding)
+    ctx_options |= (ctx_option_t)SSL_OP_TLSEXT_PADDING;
+
+#ifdef SSL_OP_TLSEXT_PADDING_SUPER
+  if(ssl_config->tlsext_padding_super)
+    ctx_options |= (ctx_option_t)SSL_OP_TLSEXT_PADDING_SUPER;
+#endif
+
   switch(ssl_version_min) {
   case CURL_SSLVERSION_SSLv2:
   case CURL_SSLVERSION_SSLv3:
