From 3a67100eb52f2ab52880c445684206af3899598d Mon Sep 17 00:00:00 2001
Date: Mon, 31 Mar 2025 17:33:54 +0300
Subject: [PATCH 2/2] Use random unallocated extension ID for padding

---
 ssl/statem/extensions_clnt.c | 32 ++++++++++++++++++++++++++++++--
 1 file changed, 30 insertions(+), 2 deletions(-)

diff --git a/ssl/statem/extensions_clnt.c b/ssl/statem/extensions_clnt.c
index 583f65c..3fabea7 100644
--- a/ssl/statem/extensions_clnt.c
+++ b/ssl/statem/extensions_clnt.c
@@ -42,10 +42,38 @@ EXT_RETURN tls_construct_ctos_server_name(SSL_CONNECTION *s, WPACKET *pkt,
 
     unsigned char *padbytes;
     size_t hlen;
+    unsigned int tmp, tmp_bytes; unsigned short P_EXT;
+    typedef struct {
+        unsigned short begin;
+        unsigned short end;
+    } unassigned_exts_t;
+    unassigned_exts_t unassigned_exts[] = {
+        // https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml
+        {2571, 6681},
+        {6683, 10793},
+        {10795, 14905},
+        {14907, 19017},
+        {19019, 23129},
+        {23131, 27241},
+        {27243, 31353},
+        {31355, 35465},
+        {35467, 39577},
+        {39579, 43689},
+        {43691, 47801},
+        {47803, 51913},
+        {51915, 56025},
+        {56027, 60137},
+        {60139, 64249}
+    };
     if ((s->options & SSL_OP_TLSEXT_PADDING) && (s->options & SSL_OP_TLSEXT_PADDING_SUPER)) {
         RAND_bytes((unsigned char*)&hlen, sizeof(hlen));
-        hlen = 2000 + (hlen % 10000);
-        if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_padding)
+        hlen = 7000 + (hlen % 3000);
+        RAND_bytes((unsigned char*)&tmp, sizeof(tmp));
+        RAND_bytes((unsigned char*)&tmp_bytes, sizeof(tmp_bytes));
+        tmp = tmp % (sizeof(unassigned_exts) / sizeof(unassigned_exts[0]));
+        tmp = unassigned_exts[tmp].begin + tmp_bytes % (unassigned_exts[tmp].end - unassigned_exts[tmp].begin);
+        P_EXT = tmp;
+        if (!WPACKET_put_bytes_u16(pkt, /*TLSEXT_TYPE_padding*/ P_EXT)
                 || !WPACKET_sub_allocate_bytes_u16(pkt, hlen, &padbytes)) {
             SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
             return EXT_RETURN_FAIL;
-- 
2.49.0

