From 9fbabfd53559338736a32a206e5a1da6f2265340 Mon Sep 17 00:00:00 2001
Date: Wed, 7 Jun 2023 23:34:35 +0300
Subject: [PATCH 1/4] Use Point-to-Point links only on Windows Vista and
 earlier

---
 src/openvpn/options.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/openvpn/options.c b/src/openvpn/options.c
index ce7bd0a..32b820f 100644
--- a/src/openvpn/options.c
+++ b/src/openvpn/options.c
@@ -3408,7 +3408,7 @@ options_postprocess_mutate_invariant(struct options *options)
         /* delay may only be necessary when we perform DHCP handshake */
         const bool dhcp = (options->tuntap_options.ip_win32_type == IPW32_SET_DHCP_MASQ)
                           || (options->tuntap_options.ip_win32_type == IPW32_SET_ADAPTIVE);
-        if ((options->mode == MODE_POINT_TO_POINT) && dhcp)
+        if ((options->mode == MODE_POINT_TO_POINT) && dhcp && (win32_version_info() <= WIN_VISTA))
         {
             options->route_delay_defined = true;
             options->route_delay = 5; /* Vista sometimes has a race without this */
-- 
2.40.1

