diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3240be78..16d5880d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -96,9 +96,9 @@ else ()
     endif ()
 endif ()
 
-find_program(PYTHON NAMES python3 python)
+find_package(Python COMPONENTS Interpreter)
 execute_process(
-    COMMAND ${PYTHON} ${CMAKE_CURRENT_SOURCE_DIR}/contrib/cmake/parse-version.m4.py ${CMAKE_CURRENT_SOURCE_DIR}/version.m4
+    COMMAND ${Python_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/contrib/cmake/parse-version.m4.py ${CMAKE_CURRENT_SOURCE_DIR}/version.m4
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
     )
 include(${CMAKE_CURRENT_BINARY_DIR}/version.cmake)
@@ -286,7 +286,7 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR} src/compat include)
 
 add_custom_command(
     OUTPUT always_rebuild config-version.h
-    COMMAND ${PYTHON} ${CMAKE_CURRENT_SOURCE_DIR}/contrib/cmake/git-version.py
+    COMMAND ${Python_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/contrib/cmake/git-version.py
     )
 set(HAVE_CONFIG_VERSION_H YES)
 
@@ -737,3 +737,7 @@ if (BUILD_TESTING)
     endif ()
 
 endif (BUILD_TESTING)
+
+install(TARGETS openvpn
+   RUNTIME
+)
