From c5f65c439b3eb1e7ceefd051bcfe68fb8591fb55 Mon Sep 17 00:00:00 2001
Date: Thu, 8 Jun 2023 02:45:48 +0300
Subject: [PATCH 1/3] vc.mak

---
 src/vc.mak | 62 ++++++++++++++++++++++++++----------------------------
 1 file changed, 30 insertions(+), 32 deletions(-)

diff --git a/src/vc.mak b/src/vc.mak
index 500135d..bdc344e 100644
--- a/src/vc.mak
+++ b/src/vc.mak
@@ -1,4 +1,4 @@
-# vc.mak by Michal Trojnara 1998-2022
+# vc.mak by Michal Trojnara 1998-2021
 # with help of David Gillingham <dgillingham@gmail.com>
 # with help of Pierre Delaage <delaage.pierre@free.fr>
 
@@ -11,10 +11,8 @@
 !IF [ml64.exe /help >NUL 2>&1]
 TARGET=win32
 SSLLIBS=libcrypto.lib libssl.lib
-# or change libraries for OpenSSL older than 1.1.0
-#SSLLIBS=libeay32.lib ssleay32.lib
 !ELSE
-TARGET=win64
+TARGET=$(TARGET_ARCH)
 SSLLIBS=libcrypto.lib libssl.lib
 !ENDIF
 !MESSAGE Detected target: $(TARGET)
@@ -25,9 +23,9 @@ SSLLIBS=libcrypto.lib libssl.lib
 # http://www.slproweb.com/products/Win32OpenSSL.html
 #SSLDIR=C:\OpenSSL-Win32
 # or compile one yourself
-SSLDIR=\devel\$(TARGET)\openssl
+#SSLDIR=\devel\$(TARGET)\openssl
 # or simply install with "nmake -f ms\ntdll.mak install"
-#SSLDIR=\usr\local\ssl
+SSLDIR=$(OPENSSL_HOME)
 
 INCDIR=$(SSLDIR)\include
 LIBDIR=$(SSLDIR)\lib
@@ -39,11 +37,11 @@ BINROOT=..\bin
 BIN=$(BINROOT)\$(TARGET)
 
 SHAREDOBJS=$(OBJ)\stunnel.obj $(OBJ)\ssl.obj $(OBJ)\ctx.obj \
-	$(OBJ)\verify.obj $(OBJ)\file.obj $(OBJ)\client.obj \
-	$(OBJ)\protocol.obj $(OBJ)\sthreads.obj $(OBJ)\log.obj \
-	$(OBJ)\options.obj $(OBJ)\network.obj $(OBJ)\resolver.obj \
-	$(OBJ)\str.obj $(OBJ)\tls.obj $(OBJ)\fd.obj $(OBJ)\dhparam.obj \
-	$(OBJ)\cron.obj
+    $(OBJ)\verify.obj $(OBJ)\file.obj $(OBJ)\client.obj \
+    $(OBJ)\protocol.obj $(OBJ)\sthreads.obj $(OBJ)\log.obj \
+    $(OBJ)\options.obj $(OBJ)\network.obj $(OBJ)\resolver.obj \
+    $(OBJ)\str.obj $(OBJ)\tls.obj $(OBJ)\fd.obj $(OBJ)\dhparam.obj \
+    $(OBJ)\cron.obj
 GUIOBJS=$(OBJ)\ui_win_gui.obj $(OBJ)\resources.res
 CLIOBJS=$(OBJ)\ui_win_cli.obj
 
@@ -61,45 +59,45 @@ CLILIBS=
 #	/LIBPATH:"$(LIBDIR)\VC\static" libeay32MD.lib ssleay32MD.lib
 
 {$(SRC)\}.c{$(OBJ)\}.obj:
-	$(CC) $(CFLAGS) -Fo$@ -c $<
+    $(CC) $(CFLAGS) -Fo$@ -c $<
 
 {$(SRC)\}.rc{$(OBJ)\}.res:
-	$(RC) -fo$@ -r $<
+    $(RC) -fo$@ -r $<
 
 all: build
 
 build: makedirs $(BIN)\stunnel.exe $(BIN)\tstunnel.exe
 
 clean:
-	-@ del $(SHAREDOBJS) >NUL 2>&1
-	-@ del $(GUIOBJS) >NUL 2>&1
-	-@ del $(CLIOBJS) >NUL 2>&1
+    -@ del $(SHAREDOBJS) >NUL 2>&1
+    -@ del $(GUIOBJS) >NUL 2>&1
+    -@ del $(CLIOBJS) >NUL 2>&1
 #	-@ del *.manifest >NUL 2>&1
-	-@ del $(BIN)\stunnel.exe >NUL 2>&1
-	-@ del $(BIN)\stunnel.exe.manifest >NUL 2>&1
-	-@ del $(BIN)\tstunnel.exe >NUL 2>&1
-	-@ del $(BIN)\tstunnel.exe.manifest >NUL 2>&1
-	-@ rmdir $(OBJ) >NUL 2>&1
-	-@ rmdir $(BIN) >NUL 2>&1
+    -@ del $(BIN)\stunnel.exe >NUL 2>&1
+    -@ del $(BIN)\stunnel.exe.manifest >NUL 2>&1
+    -@ del $(BIN)\tstunnel.exe >NUL 2>&1
+    -@ del $(BIN)\tstunnel.exe.manifest >NUL 2>&1
+    -@ rmdir $(OBJ) >NUL 2>&1
+    -@ rmdir $(BIN) >NUL 2>&1
 
 makedirs:
-	-@ IF NOT EXIST $(OBJROOT) mkdir $(OBJROOT) >NUL 2>&1
-	-@ IF NOT EXIST $(OBJ) mkdir $(OBJ) >NUL 2>&1
-	-@ IF NOT EXIST $(BINROOT) mkdir $(BINROOT) >NUL 2>&1
-	-@ IF NOT EXIST $(BIN) mkdir $(BIN) >NUL 2>&1
+    -@ IF NOT EXIST $(OBJROOT) mkdir $(OBJROOT) >NUL 2>&1
+    -@ IF NOT EXIST $(OBJ) mkdir $(OBJ) >NUL 2>&1
+    -@ IF NOT EXIST $(BINROOT) mkdir $(BINROOT) >NUL 2>&1
+    -@ IF NOT EXIST $(BIN) mkdir $(BIN) >NUL 2>&1
 
 $(SHAREDOBJS): *.h vc.mak
 $(GUIOBJS): *.h vc.mak
 $(CLIOBJS): *.h vc.mak
 
 $(BIN)\stunnel.exe: $(SHAREDOBJS) $(GUIOBJS)
-	$(LINK) $(LDFLAGS) $(SHAREDLIBS) $(GUILIBS) /LIBPATH:"$(LIBDIR)" $(SSLLIBS) /OUT:$@ $**
-	IF EXIST $@.manifest \
-		mt -nologo -manifest $@.manifest -outputresource:$@;1
+    $(LINK) $(LDFLAGS) $(SHAREDLIBS) $(GUILIBS) /LIBPATH:"$(LIBDIR)" $(SSLLIBS) /OUT:$@ $**
+    IF EXIST $@.manifest \
+        mt -nologo -manifest $@.manifest -outputresource:$@;1
 
 $(BIN)\tstunnel.exe: $(SHAREDOBJS) $(CLIOBJS)
-	$(LINK) $(LDFLAGS) $(SHAREDLIBS) $(CLILIBS) /LIBPATH:"$(LIBDIR)" $(SSLLIBS) /OUT:$@ $**
-	IF EXIST $@.manifest \
-		mt -nologo -manifest $@.manifest -outputresource:$@;1
+    $(LINK) $(LDFLAGS) $(SHAREDLIBS) $(CLILIBS) /LIBPATH:"$(LIBDIR)" $(SSLLIBS) /OUT:$@ $**
+    IF EXIST $@.manifest \
+        mt -nologo -manifest $@.manifest -outputresource:$@;1
 
 # end of vc.mak
-- 
2.40.1.windows.1

