From 8d2e7f7d4044cd27850a5f630cddc8f20c421ab0 Mon Sep 17 00:00:00 2001
Date: Thu, 8 Jun 2023 02:47:42 +0300
Subject: [PATCH 3/3] Add support for TLSEXT_PADDING_SUPER

---
 src/options.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/options.c b/src/options.c
index 526ddb3..9888588 100644
--- a/src/options.c
+++ b/src/options.c
@@ -263,6 +263,9 @@ static const SSL_OPTION ssl_opts[] = {
 #ifdef SSL_OP_TLSEXT_PADDING
     {"TLSEXT_PADDING", SSL_OP_TLSEXT_PADDING},
 #endif
+#ifdef SSL_OP_TLSEXT_PADDING_SUPER
+    {"TLSEXT_PADDING_SUPER", SSL_OP_TLSEXT_PADDING_SUPER},
+#endif
 #ifdef SSL_OP_TLS_ROLLBACK_BUG
     {"TLS_ROLLBACK_BUG", SSL_OP_TLS_ROLLBACK_BUG},
 #endif
-- 
2.40.1.windows.1

