From c22d8496fe7c2e15f4288b64a945f2bb0d631ad1 Mon Sep 17 00:00:00 2001
Date: Thu, 8 Jun 2023 02:56:19 +0300
Subject: [PATCH 1/2] HPKE build fix

---
 crypto/hpke/hpke.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/crypto/hpke/hpke.c b/crypto/hpke/hpke.c
index 653b3a2..d936e1c 100644
--- a/crypto/hpke/hpke.c
+++ b/crypto/hpke/hpke.c
@@ -16,7 +16,7 @@
 #include <stdint.h>
 #include <string.h>
 
-#include <arpa/inet.h>
+//#include <arpa/inet.h>
 
 #include <openssl/ssl.h>
 #include <openssl/rand.h>
-- 
2.40.1.windows.1

